astra/src/main/java/com/slack/astra/blobfs/BlobStore.java

```diff
@@ -195,6 +197,28 @@ public List<String> listFiles(String prefix) {
     return filesList;
   }
 
+  public Map<String, Long> listFilesWithSize(String prefix) {
+    assert prefix != null && !prefix.isEmpty();
+
+    ListObjectsV2Request listRequest = builder().bucket(bucketName).prefix(prefix).build();
+    ListObjectsV2Publisher asyncPaginatedListResponse =
+        s3AsyncClient.listObjectsV2Paginator(listRequest);
+
+    Map<String, Long> filesListWithSize = new HashMap<>();
+    try {
+      asyncPaginatedListResponse
+          .subscribe(
+              listResponse ->
+                  listResponse
+                      .contents()
+                      .forEach(s3Object -> filesListWithSize.put(s3Object.key(), s3Object.size())))
+          .get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+    return filesListWithSize;
```

Comment:
you could extract a helper method that would be used by both listFiles... methods that takes a prefix and a Consumer. Then this could look like the following.

Also, the block passed to subscribe could be called in multiple threads, so this should use a storage class that is safe wrt concurrent modifications.

```suggestion
    Map<String, Long> filesWithSize = new ConcurrentHashMap<>();
    listFilesAndDo(prefix, s3Object -> filesListWithSize.put(s3Object.key(), s3Object.size()));
    return filesWithSize;
```